pkg/repository/postgresql/postgresql.go

```diff
@@ -0,0 +1,141 @@
+func NewContactsRepositoryPostgreSQL(db *sql.DB) *ContactsRepositoryPostgreSQL {
+	return &ContactsRepositoryPostgreSQL{
+		db: db,
+	}
+}
+
+func (r *ContactsRepositoryPostgreSQL) Save(contact model.Contact) (model.Contact, error) {
+	query := "INSERT INTO contacts(firstname,lastname,phone,email) VALUES($1,$2,$3,$4) returning id;"
+	err := r.db.QueryRow(query, contact.FirstName, contact.LastName, contact.Phone, contact.Email).Scan(&contact.ID)
+
+	if err != nil {
+		if strings.Contains(err.Error(), "duplicate key value violates unique constraint \"contacts_phone_idx\"") {
+			return model.Contact{}, fmt.Errorf("contact with phone %q already exists", contact.Phone)
+		}
+
+		if strings.Contains(err.Error(), "duplicate key value violates unique constraint \"contacts_email_idx\"") {
+			return model.Contact{}, fmt.Errorf("contact with email %q already exists", contact.Email)
+		}
+
+		return model.Contact{}, err
+	}
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryPostgreSQL) ListAll() (contacts []model.Contact, err error) {
+	query := "SELECT id, firstname, lastname, email, phone FROM contacts;"
+	rows, err := r.db.Query(query)
+	if err != nil {
+		return contacts, err
+	}
+
+	defer func() {
+		_ = rows.Close()
+	}()
+
+	for rows.Next() {
+		var contact model.Contact
+		err = rows.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Email, &contact.Phone)
+		if err != nil {
+			return nil, err
+		}
+		contacts = append(contacts, contact)
+	}
+
+	return contacts, nil
+}
+
+func (r *ContactsRepositoryPostgreSQL) GetByID(id uint) (contact model.Contact, err error) {
+	query := "SELECT id, firstname, lastname, email, phone FROM contacts WHERE id = $1;"
+	row := r.db.QueryRow(query, id)
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Email, &contact.Phone)
+
+	if err != nil {
+		if err != sql.ErrNoRows {
+			return model.Contact{}, err
+		}
+
+		return model.Contact{}, fmt.Errorf("record not found")
+	}
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryPostgreSQL) GetByPhone(phone string) (contact model.Contact, err error) {
+	query := "SELECT id, firstname, lastname, email, phone FROM contacts WHERE phone = $1;"
+	row := r.db.QueryRow(query, phone)
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Email, &contact.Phone)
+
+	if err != nil {
+		if err != sql.ErrNoRows {
+			return model.Contact{}, err
+		}
+
+		return model.Contact{}, fmt.Errorf("record not found")
+	}
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryPostgreSQL) GetByEmail(email string) (contact model.Contact, err error) {
+	query := "SELECT id, firstname, lastname, email, phone FROM contacts WHERE email = $1;"
+	row := r.db.QueryRow(query, email)
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Email, &contact.Phone)
+
+	if err != nil {
+		if err != sql.ErrNoRows {
+			return model.Contact{}, err
+		}
+
+		return model.Contact{}, fmt.Errorf("record not found")
+	}
```

Comment:
```
if err == sql.ErrNoRows {
  return model.Contact{}, fmt.Errorf("record not found")
} else if err != nil {
    return model.Contact{}, err
}
```

Reply:
In your case if `err == nil` it will check `err == sql.ErrNoRows` and after that also `err != nil`, in my case only `err != nil`.
"Экономия на спичках", but still :)

Maybe it should look something like? If we have any error - we will check if-else cases inside of `if err != nil` and in case of an unexcepted type of error we will return `return model.Contact{}, err`

```
if err != nil {
	if err == sql.ErrNoRows {
		return model.Contact{}, fmt.Errorf("record not found")
	}

	return model.Contact{}, err
}
```
What u think about it?

Reply:
as a plain struct this can be used 
```go
if err == sql.ErrNoErrors{
    return  // no records found
}
if err != nil {
    return // real error
}
//proceed logic
```

Reply:
Yeah, but `if there is no error`, then `2 checks will be made instead of 1`.
If it is all wrapped in a _common if_ it will check only once and will go further.
Most of the responses will be without errors and will make 2 checks instead of 1 in your case, and only if we have an error we will check it once instead of 2 in my case. But by default, we will have more responses without errors than with errors.
I understand that writing 2 ifs as a plain might be more readable, but how can it come up in terms of performance in some cases?